v2/kafka-to-bigquery/src/main/java/com/google/cloud/teleport/v2/transforms/BigQueryDynamicDestination.java

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright (C) 2024 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package com.google.cloud.teleport.v2.transforms;
+
+import com.google.api.services.bigquery.model.TableRow;
+import com.google.api.services.bigquery.model.TableSchema;
+import com.google.cloud.teleport.v2.kafka.coders.GenericRecordCoder;
+import org.apache.avro.generic.GenericRecord;
+import org.apache.beam.sdk.coders.Coder;
+import org.apache.beam.sdk.extensions.avro.schemas.utils.AvroUtils;
+import org.apache.beam.sdk.io.gcp.bigquery.BigQueryUtils;
+import org.apache.beam.sdk.io.gcp.bigquery.DynamicDestinations;
+import org.apache.beam.sdk.io.gcp.bigquery.TableDestination;
+import org.apache.beam.sdk.values.KV;
+import org.apache.beam.sdk.values.ValueInSingleWindow;
+
+public class BigQueryDynamicDestination
+    extends DynamicDestinations<KV<GenericRecord, TableRow>, GenericRecord> {
+
+  private String projectName;
+
+  private String datasetName;
+
+  private String tableNamePrefix;
+
+  public static BigQueryDynamicDestination of(
+      String projectName, String datasetName, String tableNamePrefix) {
+    return new BigQueryDynamicDestination(projectName, datasetName, tableNamePrefix);
+  }
+
+  private BigQueryDynamicDestination(
+      String projectName, String datasetName, String tableNamePrefix) {
+    this.projectName = projectName;
+    this.datasetName = datasetName;
+    this.tableNamePrefix = tableNamePrefix;
+  }
+
+  @Override
+  public GenericRecord getDestination(ValueInSingleWindow<KV<GenericRecord, TableRow>> element) {
+    return element.getValue().getKey();
+  }
+
+  @Override
+  public TableDestination getTable(GenericRecord element) {
+    // tablename + record name (same across schemas) + schema id?
+    String bqQualifiedFullName = element.getSchema().getFullName().replace(".", "-");
+    String tableName =
+        this.tableNamePrefix + (this.tableNamePrefix == "" ? "" : "-") + bqQualifiedFullName;
+    String tableSpec = this.projectName + ":" + this.datasetName + "." + tableName;
+    return new TableDestination(tableSpec, null);
+  }
+
+  @Override
+  public TableSchema getSchema(GenericRecord element) {
+    return BigQueryUtils.toTableSchema(AvroUtils.toBeamSchema(element.getSchema()));
```

Comment:
I think when we write TableRows, this method is not required, and it can return null. Can you check if it works with nulls? If it does, that's gonna be a little more efficient (we won't do unnecessary schema conversions).

Reply:
(OK to do it later, but at least add a TODO)

Reply:
IIUC this can be null if the tables are already there, we need the schema if we want to create tables at run time. I can test the behaviour though.

Reply:
Can't return a null
```
java.lang.RuntimeException: java.lang.IllegalArgumentException: Unless create disposition is CREATE_NEVER, a schema must be specified, i.e. DynamicDestinations.getSchema() may not return null. However, create disposition is CREATE_IF_NEEDED, and  tableSpec: <tableSpec> returned null for destination %s
```

Reply:
I think for the destination table we can use CREATE_NEVER disposition, then this can be null.